Lib9c/Model/State/ShopState.cs

```diff
@@ -9,126 +10,228 @@
 namespace Nekoyume.Model.State
 {
     /// <summary>
-    /// Shop의 상태 모델이다.
-    /// 
-    /// ---- 지금의 상점의 동기화 정책.
-    /// `Sell` 액션에 대해서는 매번 직접 `Register`.
-    /// `SellCancellation` 액션에 대해서도 매번 직접 `Unregister`.
-    /// `Buy` 액션에 대해서도 매번 직접 `Unregister`.
-    /// ShopAddress의 Shop 자체에 대한 동기화는 게임 실행 시 한 번.
-    ///
-    /// ---- 추후에 예정된 이슈.
-    /// 상점의 아이템 수는 계속 증가할 것인데, 나중에는 전부를 동기화 하는 것이 무리라고 생각됨.
-    /// 상점을 단일 상태로 관리하지 않고, 1000개나 10000개 정도를 갖고 있는 단위로 채널 처럼 관리하는 것이 좋겠음.
-    /// 무작위로 접근해서 조회하도록.
-    /// 단, 이때 각 아바타의 판매 목록을 불러오는 것에 문제가 생기니, 이 목록에 접근하는 방법을 아바타의 상태에 포함해야 함.
+    /// This is a model class of shop state.
     /// </summary>
     [Serializable]
     public class ShopState : State
     {
-        public static readonly Address Address = Addresses.Shop; 
-        public readonly Dictionary<Address, List<ShopItem>> AgentProducts = new Dictionary<Address, List<ShopItem>>();
+        public static readonly Address Address = Addresses.Shop;
+
+        private readonly Dictionary<Address, List<Guid>> _agentProducts =
+            new Dictionary<Address, List<Guid>>();
+
+        private readonly Dictionary<Guid, ShopItem> _products = new Dictionary<Guid, ShopItem>();
+
+        private readonly Dictionary<ItemSubType, List<Guid>> _itemSubTypeProducts =
+            new Dictionary<ItemSubType, List<Guid>>(ItemSubTypeComparer.Instance);
+
+        public IReadOnlyDictionary<Address, List<Guid>> AgentProducts => _agentProducts;
+
+        public IReadOnlyDictionary<Guid, ShopItem> Products => _products;
+
+        public IReadOnlyDictionary<ItemSubType, List<Guid>> ItemSubTypeProducts => _itemSubTypeProducts;
 
         public ShopState() : base(Address)
         {
         }
 
-        public ShopState(Dictionary serialized)
-            : base(serialized)
+        public ShopState(Dictionary serialized) : base(serialized)
         {
-            AgentProducts = ((Dictionary)serialized["agentProducts"]).ToDictionary(
+            _agentProducts = ((Dictionary) serialized["agentProducts"]).ToDictionary(
                 kv => kv.Key.ToAddress(),
-                kv => ((List)kv.Value)
-                    .Select(d => new ShopItem((Dictionary)d))
-                    .ToList()
-            );
+                kv => ((List) kv.Value)
+                    .Select(d => d.ToGuid())
+                    .ToList());
+
+            _products = ((Dictionary) serialized["products"]).ToDictionary(
+                kv => kv.Key.ToGuid(),
+                kv => new ShopItem((Dictionary) kv.Value));
+
+            _itemSubTypeProducts = ((Dictionary) serialized["itemSubTypeProducts"]).ToDictionary(
+                kv => kv.Key.ToEnum<ItemSubType>(),
+                kv => ((List) kv.Value)
+                    .Select(value => value.ToGuid())
+                    .ToList());
         }
 
-        public ShopItem Register(Address sellerAgentAddress, ShopItem shopItem)
+        public override IValue Serialize() =>
```

Comment:
it looks `_agentProducts, _itemSubTypeProducts` is filter of `_products` right? i think serialize only `products` and can set `_agentProducts, _itemSubTypeProducts` in constructor.

Reply:
Since `_products` does not have the address of the seller `AgentState`, it must be used with `_agentProducts`.
So, if you follow the suggestion, you can create `_itemSubTypeProducts` in the constructor of `ShopState`.
I also pondered this for a while, but I wanted to write a logic that favors speed rather than capacity.
So, I tried the following tests.

```
public class ShopStateTest
{
    private readonly ITestOutputHelper _output;
    private readonly ShopState _shopState;
    private readonly Dictionary _serializedShopState;

    public ShopStateTest(ITestOutputHelper output)
    {
        _output = output;
        _shopState = new ShopState();

        var agentAddress = new PrivateKey().ToAddress();
        var avatarAddress = new PrivateKey().ToAddress();
        var weaponRow = new EquipmentItemSheet.Row();
        weaponRow.Set(new[]
        {
            "10100000", "Weapon", "0", "Normal", "0", "ATK", "1", "2", "10100000",
        });

        for (var i = 0; i < 100000; i++)
        {
            var productId = Guid.NewGuid();
            var itemUsable = new Weapon(
                weaponRow,
                Guid.NewGuid(),
                0);
            var shopItem = new ShopItem(
                avatarAddress,
                productId,
                itemUsable,
                0
            );
            _shopState.Register(agentAddress, shopItem);
        }

        _serializedShopState = (Dictionary)_shopState.Serialize();
    }

    [Fact]
    public void SerializeTime()
    {
        var sw = new Stopwatch();
        sw.Start();
        _shopState.Serialize();
        sw.Stop();
        _output.WriteLine(sw.Elapsed.ToString());
    }

    [Fact]
    public void DeserializeTime()
    {
        var sw = new Stopwatch();
        sw.Start();
        var shopState = new ShopState(_serializedShopState);
        sw.Stop();
        _output.WriteLine(sw.Elapsed.ToString());
    }
}
```

### The result of including the `_itemSubTypeProducts` in serialization as in PR
   - `SerializeTIme`: 00:00:08.6082262
   - `DeserializeTime`: 00:00:04.7373098

Next, I removed the `_itemSubTypeProducts` from serialization and wrote the logic to create it in the constructor.

```
public ShopState() : base(Address)
{
    _itemSubTypeProducts = new Dictionary<ItemSubType, List<Guid>>();
    foreach (var product in _products)
    {
        var itemSubType = product.Value.ItemUsable.ItemSubType;
        if (!_itemSubTypeProducts.ContainsKey(itemSubType))
        {
            _itemSubTypeProducts.Add(itemSubType, new List<Guid>());
        }

        _itemSubTypeProducts[itemSubType].Add(product.Value.ProductId);
    }
}

public ShopState(Dictionary serialized) : base(serialized)
{
    _agentProducts = ((Dictionary) serialized["agentProducts"]).ToDictionary(
        kv => kv.Key.ToAddress(),
        kv => ((List) kv.Value)
            .Select(d => d.ToGuid())
            .ToList());

    _products = ((Dictionary) serialized["products"]).ToDictionary(
        kv => kv.Key.ToGuid(),
        kv => new ShopItem((Dictionary) kv.Value));

    // _itemSubTypeProducts = ((Dictionary) serialized["itemSubTypeProducts"]).ToDictionary(
    //     kv => kv.Key.ToEnum<ItemSubType>(),
    //     kv => ((List) kv.Value)
    //         .Select(value => value.ToGuid())
    //         .ToList());
}

public override IValue Serialize() =>
    new Dictionary(new Dictionary<IKey, IValue>
    {
        [(Text) "agentProducts"] = new Dictionary(
            _agentProducts.Select(kv =>
                new KeyValuePair<IKey, IValue>(
                    (Binary) kv.Key.Serialize(),
                    new List(kv.Value.Select(i => i.Serialize()))))),
        [(Text) "products"] = new Dictionary(
            _products.Select(kv =>
                new KeyValuePair<IKey, IValue>(
                    (Binary) kv.Key.Serialize(),
                    kv.Value.Serialize()))),
        // [(Text) "itemSubTypeProducts"] = new Dictionary(
        //     _itemSubTypeProducts.Select(kv =>
        //         new KeyValuePair<IKey, IValue>(
        //             (Text) kv.Key.Serialize(),
        //             new List(kv.Value.Select(i => i.Serialize()))))),
    }.Union((Dictionary) base.Serialize()));
```

### As in the suggestion, the result of removing `_itemSubTypeProducts` from serialization and including it in the constructor of `ShopState`
   - `SerializeTIme`: 00:00:07.9155835
   - `DeserializeTime`: 00:00:01.3615111

### Summary

Although the number of tests is small, I have seen a huge improvement in `DeserializeTime`.

Reply:
In my opinion, adding the fields required for indexing (ex:`itemSubType, agentAddress`) to the `ShopItem`. you can serialize only `_products`.